SQL RAG AGENT

In [1]:
import os
from dotenv import load_dotenv 

load_dotenv()

True

In [2]:
from langchain_text_splitters import CharacterTextSplitter
from glob import glob

text_splitter = CharacterTextSplitter(
    separator="###",
    chunk_size=1000,
    chunk_overlap=200,
    length_function=len,
    is_separator_regex=False,
)

texts = []
for file_path in glob("docs/*.txt"): # loading all text files in docs folder
    with open(file_path, "r") as f:
        doc = f.read()
    
    file_chunks = text_splitter.create_documents(
        [doc], 
        metadatas=[{"source": file_path}]
    )
    texts.extend(file_chunks)

if texts:
    print(texts[0])

/home/ajay/miniconda3/envs/agents/lib/python3.12/site-packages/tqdm/auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html
  from .autonotebook import tqdm as notebook_tqdm
Created a chunk of size 1011, which is longer than the specified 1000
Created a chunk of size 1187, which is longer than the specified 1000
Created a chunk of size 1012, which is longer than the specified 1000


page_content='Customers Table

Represents individual bank customers.
Primary key: customer_id (uuid)

Important fields
full_name: customer’s legal name
kyc_status: regulatory verification status
Allowed values: PENDING, VERIFIED

Lifecycle
Soft deleted using deleted_at
created_at and updated_at are system-managed

Relationships
One customer can have:
multiple accounts
multiple loans

CREATE TABLE public.customers (
    customer_id uuid DEFAULT gen_random_uuid() NOT NULL,
    full_name text NOT NULL,
    kyc_status text,
    created_at timestamp without time zone DEFAULT now() NOT NULL,
    updated_at timestamp without time zone DEFAULT now() NOT NULL,
    deleted_at timestamp without time zone,
    CONSTRAINT customers_kyc_status_check CHECK ((kyc_status = ANY (ARRAY['PENDING'::text, 'VERIFIED'::text])))
);' metadata={'source': 'docs/schema.txt'}


In [3]:
texts[-1].page_content

"Querying Guidance for LLMs\n\nWhen generating SQL:\n- Always consider deleted_at IS NULL unless explicitly querying historical data\n- Currency is INR (Indian Rupee), respond as INR, Rs, or ₹ when showing a currency value (example: INR 10, Rs.10, ₹10)\nJoin paths:\nCustomers → Accounts via customer_id\nCustomers → Loans via customer_id\nLoans → Repayments via loan_id\nBranches link to Accounts and Loans via branch_id\n\nDo not assume:\none account per customer\none loan per customer\nfull repayment in a single installment\n\n### Common Analytical Intent Mapping\n\nCustomer financial exposure → accounts + loans\nBranch performance → aggregate accounts and loans by branch\nLoan health → loan status + unpaid repayments\nRegulatory checks → customers with kyc_status = 'PENDING'\nDelinquency analysis → repayments where paid_date IS NULL AND due_date < CURRENT_DATE"

In [4]:
texts

[Document(metadata={'source': 'docs/schema.txt'}, page_content="Customers Table\n\nRepresents individual bank customers.\nPrimary key: customer_id (uuid)\n\nImportant fields\nfull_name: customer’s legal name\nkyc_status: regulatory verification status\nAllowed values: PENDING, VERIFIED\n\nLifecycle\nSoft deleted using deleted_at\ncreated_at and updated_at are system-managed\n\nRelationships\nOne customer can have:\nmultiple accounts\nmultiple loans\n\nCREATE TABLE public.customers (\n    customer_id uuid DEFAULT gen_random_uuid() NOT NULL,\n    full_name text NOT NULL,\n    kyc_status text,\n    created_at timestamp without time zone DEFAULT now() NOT NULL,\n    updated_at timestamp without time zone DEFAULT now() NOT NULL,\n    deleted_at timestamp without time zone,\n    CONSTRAINT customers_kyc_status_check CHECK ((kyc_status = ANY (ARRAY['PENDING'::text, 'VERIFIED'::text])))\n);"),
 Document(metadata={'source': 'docs/schema.txt'}, page_content='Branches Table\n\nRepresents physical

In [5]:
from langchain_huggingface.embeddings import HuggingFaceEmbeddings
from langchain_community.vectorstores import Chroma

embeddings = HuggingFaceEmbeddings(
    model_name="sentence-transformers/all-MiniLM-L6-v2"
)

schema_store = Chroma.from_documents(
    texts,
    embedding=embeddings,
    collection_name="schema_rag"
)

schema_retriever = schema_store.as_retriever(k=4)


In [6]:
schema_retriever.invoke("currency")

[Document(metadata={'source': 'docs/schema.txt'}, page_content="Querying Guidance for LLMs\n\nWhen generating SQL:\n- Always consider deleted_at IS NULL unless explicitly querying historical data\n- Currency is INR (Indian Rupee), respond as INR, Rs, or ₹ when showing a currency value (example: INR 10, Rs.10, ₹10)\nJoin paths:\nCustomers → Accounts via customer_id\nCustomers → Loans via customer_id\nLoans → Repayments via loan_id\nBranches link to Accounts and Loans via branch_id\n\nDo not assume:\none account per customer\none loan per customer\nfull repayment in a single installment\n\n### Common Analytical Intent Mapping\n\nCustomer financial exposure → accounts + loans\nBranch performance → aggregate accounts and loans by branch\nLoan health → loan status + unpaid repayments\nRegulatory checks → customers with kyc_status = 'PENDING'\nDelinquency analysis → repayments where paid_date IS NULL AND due_date < CURRENT_DATE"),
 Document(metadata={'source': 'docs/schema.txt'}, page_conten

In [7]:
from langchain_core.tools import tool
from sqlalchemy import create_engine, text
from typing import List, Tuple

engine = create_engine("postgresql+psycopg2://postgres:Nst%401995@localhost:5432/postgres")

def validate_sql(query: str):
    q = query.lower().strip()

    if not q.startswith("select"):
        raise ValueError("Only SELECT queries are allowed")

    forbidden = ["insert", "update", "delete", "drop", "alter", ";"]
    pass_kw = ['deleted_at']
    
    for kw in forbidden:
        if kw in q:
            raise ValueError(f"Forbidden SQL keyword detected: {kw}")

@tool
def run_sql_query(query: str) -> str:
    """Execute a validated read-only SQL query."""
    
    print("\n--- SQL GENERATED ---")
    print(query)

    # validate_sql(query) # need to solve deleted_at getting caught in forbidden words list

    with engine.connect() as conn:
        result = conn.execute(text(query))
        rows = result.fetchmany(5)
        return str(rows)


@tool
def get_table_schema(table_name: str) -> List[Tuple]:
    """Return column schema information for a table.
    Argument: table_name, example - customers
    """
    query = """
        SELECT column_name, data_type, is_nullable
        FROM information_schema.columns
        WHERE table_name = :table_name
        ORDER BY ordinal_position
    """
    with engine.connect() as conn:
        result = conn.execute(text(query), {"table_name": table_name})
        return result.fetchall()


@tool
def request_clarification(reason: str) -> str:
    """Ask the user for clarification before generating SQL."""
    return f"Clarification needed: {reason}"

    
@tool
def list_table():
    """Return names of table present in the database."""
    query = """
        SELECT table_name
        FROM information_schema.tables
        WHERE table_schema='public';
    """
    with engine.connect() as conn:
        result = conn.execute(text(query))
        return str(result.fetchall())


@tool
def fetch_relevant_schema(question: str) -> str:
    """
    Retrieve relevant database schema based on the user question.
    Always call this before generating SQL.
    """
    docs = schema_retriever.invoke(question)
    return "\n\n".join(d.page_content for d in docs)


@tool
def verify_column(table: str, column: str) -> bool:
    """Verify a column exists in a given table."""
    query = """
        SELECT 1
        FROM information_schema.columns
        WHERE table_name = :table AND column_name = :column
    """
    with engine.connect() as conn:
        result = conn.execute(text(query), {"table": table, "column": column})
        return result.first() is not None



In [8]:
from langchain_groq import ChatGroq

# "moonshotai/kimi-k2-instruct-0905"
llm = ChatGroq(
    model="meta-llama/llama-4-scout-17b-16e-instruct",
    temperature=0
)

SYSTEM_PROMPT = """
You are a PostgreSQL database assistant.

You MUST follow this exact sequence:
1. Call fetch_relevant_schema
2. Read schema carefully
3. Decide required tables and columns
4. Generate a SELECT-only SQL query
5. Execute it using run_sql_query
6. Answer in plain English

Before writing SQL:
- List tables you will use
- List columns you will use
- Verify each exists in schema

Rules:
- Never generate SQL before schema is retrieved
- Never guess table or column names
- Never modify data
- If schema is insufficient, ask for clarification
"""

# SYSTEM_PROMPT = """
# You are a database assistant.

# You will be given:
# - Relevant database schema context
# - A user question

# Steps:
# 1. Understand schema
# 2. Generate correct PostgreSQL SELECT query
# 3. Use run_sql tool
# 4. If query fails, fix and retry
# 5. Answer clearly

# Rules:
# - Never modify data
# - Never guess columns not in schema
# """


In [9]:
from langchain.agents import create_agent

agent = create_agent(
    llm,
    tools=[
        fetch_relevant_schema,
        get_table_schema,
        list_table,
        run_sql_query,
        request_clarification,
        verify_column
    ],
    system_prompt=SYSTEM_PROMPT,
)

In [10]:
question = "I need info of loans which were disbursed between jan 2023 and june 2023"

In [11]:
from IPython.display import display, Markdown

schema_context = schema_retriever.invoke(question)
context_text = "\n\n".join(d.page_content for d in schema_context)

response = agent.invoke({
    "messages": [
        {"role": "system", "content": SYSTEM_PROMPT},
        {"role": "system", "content": f"Schema:\n{context_text}"},
        {"role": "user", "content": question},
    ],
    # "recursion_limit": 10,
})

# agent.invoke returns a structured dict - extract text
display(Markdown(response["messages"][-1].content))


--- SQL GENERATED ---
SELECT loan_id, customer_id, branch_id, loan_type, principal, interest_rate, tenure_months, disbursed_at, status FROM loans WHERE disbursed_at >= '2023-01-01' AND disbursed_at <= '2023-06-30' AND deleted_at IS NULL


Here are the loans disbursed between January 2023 and June 2023:

- Loan ID: 1aec7993-2abd-46d4-9ae4-f1952b9ed951
  - Customer ID: ec5bf63f-2a89-460f-a8ea-5fe2674503b2
  - Branch ID: dc59f73e-d87a-486f-8f78-9dddff16106d
  - Loan Type: HOME
  - Principal: ₹1945675.00
  - Interest Rate: 13.46%
  - Tenure: 36 months
  - Disbursed Date: April 7, 2023
  - Status: NPA

- Loan ID: 70979651-1dff-46d5-95dc-c55557f6a559
  - Customer ID: 3024a891-5f3b-45a1-9c1e-e978aeb31601
  - Branch ID: 5384586d-6e16-46a7-93e2-aecd70687f04
  - Loan Type: AUTO
  - Principal: ₹791717.00
  - Interest Rate: 16.84%
  - Tenure: 60 months
  - Disbursed Date: June 25, 2023
  - Status: NPA

- Loan ID: 84954583-a5b2-45c2-9de0-10a367534495
  - Customer ID: 33ea8cc9-a813-48fb-b094-a8f37267d65b
  - Branch ID: 225ba84c-cf19-4ccb-b5f9-a04f9070fa8c
  - Loan Type: AUTO
  - Principal: ₹143085.00
  - Interest Rate: 12.36%
  - Tenure: 120 months
  - Disbursed Date: January 7, 2023
  - Status: NPA

- Loan ID: 1f1b7c72-b5fc-4bf4-86f3-f84934514e47
  - Customer ID: 60d47bc4-1d9c-49af-990f-834c70e8986e
  - Branch ID: 225ba84c-cf19-4ccb-b5f9-a04f9070fa8c
  - Loan Type: PERSONAL
  - Principal: ₹1448575.00
  - Interest Rate: 11.26%
  - Tenure: 36 months
  - Disbursed Date: February 9, 2023
  - Status: ACTIVE

- Loan ID: ff4147e9-feea-4f84-bd46-937cbb0bb4bf
  - Customer ID: 93c15b25-17e6-4f0f-bfe3-99626dff6045
  - Branch ID: 411285e5-25a5-40f8-bca3-dc28ad9cde62
  - Loan Type: HOME
  - Principal: ₹101223.00
  - Interest Rate: 10.45%
  - Tenure: 60 months
  - Disbursed Date: June 5, 2023
  - Status: ACTIVE

| Loan ID                              | Customer ID                          | Branch ID                            | Loan Type | Principal    | Interest Rate (%) | Tenure (Months) | Disbursed At | Status |
| ------------------------------------ | ------------------------------------ | ------------------------------------ | --------- | ------------ | ----------------- | --------------- | ------------ | ------ |
| 1aec7993-2abd-46d4-9ae4-f1952b9ed951 | ec5bf63f-2a89-460f-a8ea-5fe2674503b2 | dc59f73e-d87a-486f-8f78-9dddff16106d | HOME      | 1,945,675.00 | 13.46             | 36              | 2023-04-07   | NPA    |
| 70979651-1dff-46d5-95dc-c55557f6a559 | 3024a891-5f3b-45a1-9c1e-e978aeb31601 | 5384586d-6e16-46a7-93e2-aecd70687f04 | AUTO      | 791,717.00   | 16.84             | 60              | 2023-06-25   | NPA    |
| 84954583-a5b2-45c2-9de0-10a367534495 | 33ea8cc9-a813-48fb-b094-a8f37267d65b | 225ba84c-cf19-4ccb-b5f9-a04f9070fa8c | AUTO      | 143,085.00   | 12.36             | 120             | 2023-01-07   | NPA    |
| 1f1b7c72-b5fc-4bf4-86f3-f84934514e47 | 60d47bc4-1d9c-49af-990f-834c70e8986e | 225ba84c-cf19-4ccb-b5f9-a04f9070fa8c | PERSONAL  | 1,448,575.00 | 11.26             | 36              | 2023-02-09   | ACTIVE |
| ff4147e9-feea-4f84-bd46-937cbb0bb4bf | 93c15b25-17e6-4f0f-bfe3-99626dff6045 | 411285e5-25a5-40f8-bca3-dc28ad9cde62 | HOME      | 101,223.00   | 10.45             | 60              | 2023-06-05   | ACTIVE |



In [12]:
for step in agent.stream(
    {"messages": [{"role": "user", "content": question}]},
    stream_mode="values",
):
    step["messages"][-1].pretty_print()

================================ Human Message =================================

I need info of loans which were disbursed between jan 2023 and june 2023
================================== Ai Message ==================================

[][]


In [13]:
for step in agent.stream(
    {"messages": [{"role": "user", "content": "just give me the names of all tables"}]},
    stream_mode="values",
):
    step["messages"][-1].pretty_print()

================================ Human Message =================================

just give me the names of all tables
================================== Ai Message ==================================
Tool Calls:
  list_table (cnzyfecnx)
 Call ID: cnzyfecnx
  Args:
================================= Tool Message =================================
Name: list_table

[('branches',), ('customers',), ('accounts',), ('loans',), ('repayments',)]
================================== Ai Message ==================================

The tables present in the database are: branches, customers, accounts, loans, repayments.


In [14]:
def streamer_agent(question):
    for step in agent.stream(
        {"messages": [{"role": "user", "content": question}]},
        stream_mode="values",
    ):
        yield step["messages"][-1].pretty_print()

In [15]:
[msg for msg in streamer_agent("How financially exposed is the bank to each customer?")]

================================ Human Message =================================

How financially exposed is the bank to each customer?
================================== Ai Message ==================================

## Step 1: Call fetch_relevant_schema
First, I need to retrieve the relevant database schema based on the user's question to understand the structure of the data.
Tool Calls:
  fetch_relevant_schema (3rd81h699)
 Call ID: 3rd81h699
  Args:
    question: How financially exposed is the bank to each customer?
================================= Tool Message =================================
Name: fetch_relevant_schema

Querying Guidance for LLMs

When generating SQL:
- Always consider deleted_at IS NULL unless explicitly querying historical data
- Currency is INR (Indian Rupee), respond as INR, Rs, or ₹ when showing a currency value (example: INR 10, Rs.10, ₹10)
Join paths:
Customers → Accounts via customer_id
Customers → Loans via customer_id
Loans → Repayments via loan_id
Bra

[None, None, None, None, None, None, None, None]

In [16]:
[msg for msg in streamer_agent("whats the sum of outstanding loan amount of all customer?")]

================================ Human Message =================================

whats the sum of outstanding loan amount of all customer?
================================== Ai Message ==================================

## Step 1: Call fetch_relevant_schema
First, I need to retrieve the relevant database schema based on the user's question to identify the necessary tables and columns.
Tool Calls:
  fetch_relevant_schema (rf1cpm8ef)
 Call ID: rf1cpm8ef
  Args:
    question: whats the sum of outstanding loan amount of all customer?
================================= Tool Message =================================
Name: fetch_relevant_schema

Loans Table

Represents loans issued to customers.
Primary key: loan_id (uuid)
Foreign keys
customer_id → customers.customer_id
branch_id → branches.branch_id

Important fields

loan_type: HOME, PERSONAL, AUTO
principal: loan amount
interest_rate: percentage value
tenure_months: loan duration
disbursed_at: loan start date
status: loan state
Allowed v

[None, None, None, None, None, None, None, None]